crates/accelerate/src/barrier_before_final_measurement.rs

```diff
@@ -0,0 +1,118 @@
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashSet;
+use pyo3::prelude::*;
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+use qiskit_circuit::circuit_instruction::ExtraInstructionAttributes;
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::imports::BARRIER;
+use qiskit_circuit::operations::{Operation, PyInstruction};
+use qiskit_circuit::packed_instruction::{PackedInstruction, PackedOperation};
+use qiskit_circuit::Qubit;
+
+static FINAL_OP_NAMES: [&str; 2] = ["measure", "barrier"];
+
+#[pyfunction]
+pub fn barrier_before_final_measurements(
+    py: Python,
+    dag: &mut DAGCircuit,
+    label: Option<String>,
+) -> PyResult<()> {
+    let final_ops: HashSet<NodeIndex> = dag
+        .op_nodes(true)
+        .filter(|node| {
+            let NodeType::Operation(ref inst) = dag.dag()[*node] else {
+                unreachable!();
+            };
+            if !FINAL_OP_NAMES.contains(&inst.op.name()) {
+                return false;
+            }
+            let is_final_op = dag.bfs_successors(*node).all(|(_, child_successors)| {
+                !child_successors.iter().any(|suc| match dag.dag()[*suc] {
+                    NodeType::Operation(ref suc_inst) => {
+                        !FINAL_OP_NAMES.contains(&suc_inst.op.name())
+                    }
+                    _ => false,
+                })
+            });
+            is_final_op
+        })
+        .collect();
+    if final_ops.is_empty() {
+        return Ok(());
+    }
+    let ordered_node_indices: Vec<NodeIndex> = dag
+        .topological_op_nodes()?
+        .filter(|node| final_ops.contains(node))
+        .collect();
+    let final_packed_ops: Vec<PackedInstruction> = ordered_node_indices
+        .into_iter()
+        .map(|node| {
+            let NodeType::Operation(ref inst) = dag.dag()[node] else {
+                unreachable!()
+            };
+            let res = inst.clone();
+            dag.remove_op_node(node);
+            res
+        })
+        .collect();
+    let new_barrier = BARRIER
+        .get_bound(py)
+        .call1((dag.num_qubits(), label.as_deref()))?;
+
+    let new_barrier_py_inst = PyInstruction {
+        qubits: dag.num_qubits() as u32,
+        clbits: 0,
+        params: 0,
+        op_name: "barrier".to_string(),
+        control_flow: false,
+        #[cfg(feature = "cache_pygates")]
+        instruction: new_barrier.clone().unbind(),
+        #[cfg(not(feature = "cache_pygates"))]
+        instruction: new_barrier.unbind(),
```

Comment:
What is the difference between doing this or simply omitting the `#[cfg(feature = "cache_pygates")]` comment and setting the instruction argument to `new_barrier.unbind()`?

Reply:
When `cache_pygates` is necessary we need to clone `new_barrier` because it's used in the `PyInstruction` here and also again in the pycache for the `PackedInstruction` that we insert into the dag. I used the cfg macros here to conditionally compile this so we only clone if the `cache_pygates` feature is enabled and otherwise skip the `clone()` where it's not needed. This would work fine if I removed the macros and just had it always do `new_barrier.clone().unbind()` regardless of whether the feature was enabled, but it was just me micro-optimizing a little as the `clone()` and `Drop` for `new_barrier` as a `Bound<PyAny>` is just python ref counting and very cheap, but I figured I could just save the overhead of that  here.